src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Windows.cs

```diff
@@ -183,16 +182,29 @@ public unsafe string GetImpersonationUserName()
         {
             CheckWriteOperations();
 
-            const int UserNameMaxLength = Interop.Kernel32.CREDUI_MAX_USERNAME_LENGTH + 1;
-            char* userName = stackalloc char[UserNameMaxLength]; // ~1K
+            const uint UserNameMaxLength = Interop.Kernel32.CREDUI_MAX_USERNAME_LENGTH + 1;
+            char* userName = stackalloc char[(int)UserNameMaxLength]; // ~1K
 
-            if (!Interop.Kernel32.GetNamedPipeHandleState(InternalHandle, IntPtr.Zero, IntPtr.Zero,
-                IntPtr.Zero, IntPtr.Zero, userName, UserNameMaxLength))
+            if (Interop.Kernel32.GetNamedPipeHandleStateW(InternalHandle, null, null, null, null, userName, UserNameMaxLength))
             {
-                throw WinIOError(Marshal.GetLastWin32Error());
+                return new string(userName);
             }
 
-            return new string(userName);
+            int error = Marshal.GetLastWin32Error();
+            if (error == Interop.Errors.ERROR_SUCCESS && Environment.Is64BitProcess)
+            {
+                // There is a known problem in Windows where if sspicli is not loaded, this function fails unexpectedly with ERROR_SUCCESS, so we need to load it and reattempt
+                // If sspicli is already loaded, this is a no-op. 
+                // We set the handle as invalid so the library won't get unloaded.
+                Interop.Kernel32.LoadLibraryEx("sspicli.dll", IntPtr.Zero, Interop.Kernel32.LOAD_LIBRARY_SEARCH_SYSTEM32)?.SetHandleAsInvalid();
```

Comment:
In what situation would this call return null?  Even if LoadLibraryEx fails and returns a null pointer, the runtime should be manufacturing a SafeHandle for that.

Reply:
The documentation's [Return Value](https://docs.microsoft.com/en-us/windows/desktop/api/libloaderapi/nf-libloaderapi-loadlibraryexa#return-value) section says:

> If the function fails, the return value is NULL. To get extended error information, call GetLastError.

And the [Parameters section](https://docs.microsoft.com/en-us/windows/desktop/api/libloaderapi/nf-libloaderapi-loadlibraryexa#parameters) says:

> If the function cannot find the module or one of its dependencies, the function fails.

Since we are using the LOAD_LIBRARY_SEARCH_SYSTEM32 value as 3rd parameter, then the function would fail and would return null if it cannot find the sspicli.dll file in the System32 folder.

So that's the case where we would get a null value.

Reply:
> If the function fails, the return value is NULL

Yes, the native function's returned pointer would be null.  Which would then be wrapped into the SafeHandle returned.

Reply:
I am going to remove the "?" as it is not necessary. Thank you for pointing out. I'll update my PR soon.

Apologies for not responding earlier, I was trying to get my Windows 7 VM to run so I could investigate the unit test failure.